benchmarks/performance-poetry/complex-poetry/src/main/java/com/squareup/benchmarks/performance/complex/poetry/PerformancePoetryComponent.kt

```diff
@@ -37,9 +45,121 @@ class PerformancePoetryComponent : ViewModel() {
+      val root = session.parent == null
+      val tracedId = NODES_TO_TRACE.firstOrNull { it == session.identifier } != null
+      if (root) {
+        // Setup stats for render pass tracking.
+        rootSessionId = session.sessionId
+        sessionRenderStats[rootSessionId] = RenderStats()
+        // Cumulative Section that has total render pass time.
+        Trace.beginSection("RenderPassTotal")
+        Trace.beginSection("RenderPass-$totalRenderPasses")
+      }
+
+      if (tracedId) {
+        // Trace section for specially tracked WorkflowIdentifiers.
+        Trace.beginSection(session.identifier.toString() + "-$totalRenderPasses")
+      }
+
+      sessionRenderStats[rootSessionId]!!.apply {
+        // Update stats for this render pass with this node.
+        val renderStateString = renderState.toString()
+        val lastState = nodeStates[session.sessionId]
+        if (lastState == null) {
+          nodesRenderedDifferent += 1
+        } else {
+          if (lastState.contentEquals(renderStateString)) {
+            nodesRenderedTheSame += 1
+          } else {
+            nodesRenderedDifferent += 1
+          }
+        }
+        nodeStates[session.sessionId] = renderStateString
+      }
+
+      return proceed(renderProps, renderState, null).also {
+        if (tracedId) {
+          Trace.endSection()
+        }
+
+        if (root) {
+          // Double end section for the 'total' and the one for this render pass.
+          Trace.endSection()
+          Trace.endSection()
+          // Print stats for this render pass.
+          val totalNodes =
+            sessionRenderStats[rootSessionId]!!.nodesRenderedDifferent +
+              sessionRenderStats[rootSessionId]!!.nodesRenderedTheSame
+          Timber.i("~~~~~~ Workflow Render Pass Stats Begin ~~~~~~")
+          Timber.i(
+            "Nodes rendered identically in this pass:" +
+              " ${sessionRenderStats[rootSessionId]!!.nodesRenderedTheSame}"
+          )
+          Timber.i(
+            "Nodes rendered differently in this pass:" +
+              " ${sessionRenderStats[rootSessionId]!!.nodesRenderedDifferent}"
+          )
+          Timber.i("Total Nodes: $totalNodes")
+          Timber.i(
+            "Fresh Rendering Ratio:" +
+              " ${
+              sessionRenderStats[rootSessionId]!!.nodesRenderedDifferent.toDouble() /
+                totalNodes.toDouble()
+              }"
+          )
+          Timber.i("~~~~~~ Workflow Render Pass Stats End ~~~~~~")
+          totalFresh += sessionRenderStats[rootSessionId]!!.nodesRenderedDifferent
+          totalStale += sessionRenderStats[rootSessionId]!!.nodesRenderedTheSame
+          Timber.i("--------- Workflow Renderings Cumulative Stats ---------")
+          Timber.i("Total Render Passes So Far: $totalRenderPasses")
+          Timber.i(
+            "Total Fresh Rendering Ratio: ${
+            totalFresh.toDouble() / (totalFresh + totalStale).toDouble()
+            }"
```

Comment:
note: beware of complex timber logs slowing down the performance which you're trying to measure.

Reply:
I thought about having 2 separate versions - one where we run to measure the 'fresh rendering ratio' and count the render passes and another where we trace the timing. WDYT?